Support pattern matching on constants

hubertp opened the pull request:
### Pull Request Description

This change adds support for matching on constants by:
1) extending parser to allow literals in patterns
2) generate branch node for literals

Related to https://www.pivotaltracker.com/story/show/182743559

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Support pattern matching on constants

This change adds support for matching on constants by:
1) extending parser to allow literals in patterns
2) generate branch node for literals

hubertp committed:
changelog

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java:
```diff
@@ -0,0 +1,52 @@
+      Object state,
+      Text target,
+      @Cached("build()") ToJavaStringNode toJavaStringNode) {
+    if (textProfile.profile(literal.equals(toJavaStringNode.execute(target))))
```
This sounds like a possible overkill, but I'm wondering if we should use the same equality as our `==` operator, i.e. one that handles Unicode normalization. OTOH maybe it makes sense for pattern matching to be 'more exact'.

hubertp replied:
Done

radeusgd reviewed (approved):
Awesome! I can't wait to use it :smiley:

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java:
```diff
@@ -0,0 +1,52 @@
+      Object state,
+      Text target,
+      @Cached("build()") ToJavaStringNode toJavaStringNode) {
+    if (textProfile.profile(literal.equals(toJavaStringNode.execute(target))))
```
Converting `Text` to `String` just to do `equals`? I guess we could benefit from a switch to [TruffleStrings](https://github.com/oracle/graal/blob/master/truffle/docs/TruffleStrings.md) then and use [EqualsNode](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/strings/TruffleString.EqualNode.html) which would compare the values without converting to `String`.

hubertp replied:
Yeah, this is the only ugly part in this PR, I admit. 
Thanks for the suggestion, I wasn't entirely sure what is _the proper way_ to deal with this.

hubertp replied:
On the other hand, we use similar technique already with `ExpectStringNode`, `ExpectTextNode` (and `ToJavaStringNode`) in various places.

hubertp replied:
So I think that with a comparison of normalized strings this is no longer an option.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java:
```diff
@@ -0,0 +1,52 @@
+  private final ConditionProfile numProfile = ConditionProfile.createCountingProfile();
+
+  LiteralBranchNode(Object literal, RootCallTarget branch) {
+    super(branch);
```
What are the actual values of `literal`? `String`, `long`, `EnsoBigInteger`? Should they be enforced? By `assert`? By multiple variants of the constructor?

hubertp replied:
done

JaroslavTulach commented on engine/runtime/src/test/scala/org/enso/compiler/test/codegen/AstToIrTest.scala:
```diff
@@ -213,6 +213,38 @@ class AstToIrTest extends CompilerTest with Inside {
+      val ir =
+        """
+          |case foo of
+          |    Cons 1 b -> a + b
```
Interesting that `Cons 1 b` works!

hubertp replied:
It parses, it doesn't necessarily work :p

hubertp replied:
And now it works

JaroslavTulach commented on test/Tests/src/Semantic/Case_Spec.enso:
```diff
@@ -88,5 +88,18 @@ spec = Test.group "Pattern Matches" <|
         case Any of
             Any -> Nothing
             _ -> Test.fail "Expected the Any constructor to match."
+    Test.specify "should be able to match on literal values" <|
```
Shouldn't the `.enso` test also check for `Cons 1 b` style?

hubertp replied:
Yeap, now it does

JaroslavTulach reviewed (approved):
Great work in a really fast time!

hubertp committed:
Fix matching on nested literals

Added tests for matching on literals that are deeply nested. Turned out
the initial setup was wrong but was able to reuse the existing logic.

Addressed the first of the comments in PR

radeusgd commented on test/Tests/src/Semantic/Case_Spec.enso:
```diff
@@ -101,5 +101,13 @@ spec = Test.group "Pattern Matches" <|
             0     -> Test.fail "Expected value to match constant."
             "foo" -> Nothing
             _     -> Test.fail "Expected value to match constant."
+    Test.specify "should be able to match on literal values nested in constructors" <|
+        value_1 = Cons 42 Nil
+        case value_1 of
+            42                    -> Test.fail "Expected value to match constant."
+            Cons (Cons 1 Nil) Nil -> Test.fail "Expected value to match constant."
+            Cons 1 Nil            -> Test.fail "Expected value to match constant."
+            Cons 42 Nil           -> Nothing
+            _                     -> Test.fail "Expected value to match constant."
```
Maybe would be good to also add a test where one field is a literal but another is a binder? i.e. `Cons 42 foo ->` or `Cons 42 _ ->` ? Just for the sake of completeness.

hubertp replied:
added

hubertp committed:
Compare normalized strings when pattern matching

Split LiteralBranchNode into comparing strings and numeric values. That
way we can ensure that LHS is a String and avoid casting.

hubertp committed:
One more test

hubertp committed:
license review
